llvm/lib/Transforms/Yk/ControlPoint.cpp

```diff
@@ -0,0 +1,343 @@
+  Builder.CreateStore(ConstantInt::get(Context, APInt(8, 0)), GVTracing);
+  if (PrintJITState) {
+    createDebugPrint(Builder, &Mod, "stop-tracing");
+  }
+  Builder.CreateBr(BBReturn);
+
+  // Create return block. Returns the unchanged IOStruct if no compiled
+  // trace was executed, otherwise return a new IOStruct which contains the
+  // changed interpreter state.
+  Builder.SetInsertPoint(BBReturn);
+  Value *IOStruct = F->getArg(1);
+  PHINode *phi = Builder.CreatePHI(IOTy, 3);
+  phi->addIncoming(IOStruct, BBHasTrace);
+  phi->addIncoming(CTResult, BBExecuteTrace);
+  phi->addIncoming(IOStruct, BBTracing);
+  phi->addIncoming(IOStruct, BBHasNoTrace);
+  phi->addIncoming(IOStruct, BBStopTracing);
+  Builder.CreateRet(phi);
+}
+
+/// Extract all live variables that need to be passed into the control point.
+std::vector<Value *> getLiveValues(DominatorTree &DT, CallInst *OldCP) {
+  std::vector<Value *> Vec;
+  Function *Func = OldCP->getFunction();
+  for (auto &BB : *Func) {
+    if (!DT.dominates(cast<Instruction>(OldCP), &BB)) {
+      for (auto &I : BB) {
+        if ((DT.dominates(&I, cast<Instruction>(OldCP))) &&
+            (!I.getType()->isVoidTy())) {
+          Vec.push_back(&I);
+        }
+      }
+    }
+  }
+  return Vec;
+}
+
+namespace {
+struct YkControlPointPass : public ModulePass {
+  static char ID;
+
+  YkControlPointPass() : ModulePass(ID) {}
+
+  StringRef getPassName() const override {
+    return "Control Point Patching Pass";
+  }
+
+  bool runOnModule(Module &M) override {
+    LLVMContext &context = M.getContext();
+
+    CallInst *OldCPCall = findControlPointCall(M);
+
+    // Replace old control point call.
+    IRBuilder<> Builder(OldCPCall);
+    // Builder.SetInsertPoint(OldCPCall);
+
+    // Get function containing the control point.
+    Function *Caller = OldCPCall->getFunction();
+
+    // Find all live variables, which become the inputs/outputs to our
+    // control point.
+    DominatorTree DT(*Caller);
+    std::vector<Value *> LiveVals = getLiveValues(DT, OldCPCall);
+    // A control point with no inputs/outputs doesn't make sense.
+    assert(LiveVals.size() > 0);
+
+    // Generate the input/output struct and the new control point function.
+    std::vector<Type *> TypeParams;
+    for (Value *V : LiveVals) {
+      TypeParams.push_back(V->getType());
+    }
+    StructType *CPReturnTy = StructType::create(TypeParams, "IOStruct");
+    FunctionType *FType = FunctionType::get(
+        CPReturnTy, {Type::getInt32Ty(context), CPReturnTy}, false);
+    Function *NF = Function::Create(FType, GlobalVariable::ExternalLinkage,
+                                    YK_NEW_CONTROL_POINT, M);
+
+    // Generate the input to the control point which is a struct containing all
+    // live variables.
+    Value *InputStruct = cast<Value>(Constant::getNullValue(CPReturnTy));
+    assert(LiveVals.size() <= static_cast<size_t>(UINT_MAX));
+    for (size_t I = 0; I < LiveVals.size(); I++) {
```

Comment:
Does C++ have an `enumerate`-esque iterator which produces values and indexes? It might be less ugly than the `static_cast` monstrosity! If we do have to keep `static_cast`, we should document why the cast is safe and/or add an `assert` to check safety.